File: mac/Skribe/Skribe/Services/TranscriptionPipeline.swift
import Foundation

@MainActor
final class TranscriptionPipeline {
    func process(fileURL: URL, appState: AppState, modelManager: ModelManager) async {
        appState.startProcessing()
        appState.fileName = fileURL.lastPathComponent

        var wavURL: URL?

        do {
            // Phase 1: Convert audio to 16kHz mono WAV
            appState.currentPhase = .preparing
            appState.progressMessage = "Convertendo audio para formato ideal..."

            wavURL = try await AudioProcessor.convertToMono16kHz(inputURL: fileURL)
            guard let wavURL else { throw SkribeError.audioConversionFailed("Falha ao converter") }

            let duration = try await AudioProcessor.getAudioDuration(url: wavURL)
            appState.audioDuration = duration
            appState.progressMessage = "Audio preparado (\(Int(duration / 60))min \(Int(duration.truncatingRemainder(dividingBy: 60)))s)"

            // Phase 2: Load audio samples and detect voice activity
            appState.currentPhase = .transcribing
            appState.progressMessage = "Carregando audio e detectando voz..."

            let audioSamples = try AudioProcessor.loadAudioAsFloatArray(url: wavURL)

            // Quick VAD pre-check: bail out early if the file is essentially silent.
            // (Real silence skipping happens inside WhisperKit via `chunkingStrategy: .vad`.)
            appState.progressMessage = "Analisando áudio..."
            let voiceSegments = VADProcessor.detectVoiceActivity(samples: audioSamples)
            let totalDuration = Double(audioSamples.count) / 16000.0
            let voiceDuration = voiceSegments.reduce(0.0) { $0 + ($1.endTime - $1.startTime) }

            if voiceDuration < 1.0 {
                throw SkribeError.audioConversionFailed("Nenhuma fala detectada no áudio")
            }

            appState.progressMessage = "Voz detectada: \(Int(voiceDuration))s de fala em \(Int(totalDuration))s totais"

            // Phase 3: Run transcription and diarization in parallel
            let capturedWavURL = wavURL
            async let whisperSegments = modelManager.transcriptionService.transcribe(audioURL: capturedWavURL)
            async let diarSegments = modelManager.diarizationService.diarize(audioSamples: audioSamples)

            appState.currentPhase = .diarizing
            appState.progressMessage = "Transcrevendo e identificando vozes em paralelo..."

            let (transcription, diarization) = try await (whisperSegments, diarSegments)

            // Phase 4: Merge transcription with speaker labels using word-level alignment
            appState.currentPhase = .merging
            appState.progressMessage = "Combinando transcricao com vozes (word-level)..."

            let merged = Self.assignSpeakersWordLevel(
                whisperSegments: transcription,
                diarSegments: diarization
            )

            let renamed = Self.renameSpeakers(segments: merged)

            // Phase 5: Translate EN → PT if needed (improved dictionary, ~5-10s for 1h audio)
            // Check if translation is needed
            let needsTranslation = renamed.first.map { TranslationService.isLikelyEnglish($0.text) } ?? false

            let final: [SkribeSegment]
            if needsTranslation {
                appState.progressMessage = "Traduzindo para portugues (dicionario expandido ~500 palavras)..."
                let translationService = ImprovedTranslationService()
                final = await translationService.translateSegments(renamed)
                appState.progressMessage = "Traducao concluida ✓ (~80% cobertura)"
            } else {
                final = renamed
            }

            appState.finishProcessing(segments: final, fileName: fileURL.lastPathComponent)

        } catch {
            appState.failProcessing(error: error.localizedDescription)
        }

        if let wavURL {
            AudioProcessor.cleanup(url: wavURL)
        }
    }

    /// Word-level speaker assignment for maximum accuracy
    private static func assignSpeakersWordLevel(
        whisperSegments: [TranscriptionService.SegmentWithWords],
        diarSegments: [DiarizationService.SpeakerSegmentResult]
    ) -> [SkribeSegment] {
        var result: [SkribeSegment] = []

        for segment in whisperSegments {
            // If segment has word-level timestamps, use them for precise alignment
            if !segment.words.isEmpty {
                // Group consecutive words by speaker
                var currentSpeaker: Int?
                var currentWords: [TranscriptionService.WordTiming] = []
                var currentStart: Double?

                for word in segment.words {
                    let wordMid = (word.start + word.end) / 2
                    let speakerId = findSpeakerAtTime(wordMid, in: diarSegments)

                    // New speaker or first word
                    if speakerId != currentSpeaker {
                        // Flush previous group
                        if !currentWords.isEmpty, let start = currentStart, let speaker = currentSpeaker {
                            let text = currentWords.map { $0.word }.joined()
                            let end = currentWords.last?.end ?? start
                            result.append(SkribeSegment(
                                start: start,
                                end: end,
                                speaker: "SPEAKER_\(speaker)",
                                text: text.trimmingCharacters(in: .whitespaces)
                            ))
                        }

                        // Start new group
                        currentSpeaker = speakerId
                        currentWords = [word]
                        currentStart = word.start
                    } else {
                        currentWords.append(word)
                    }
                }

                // Flush final group
                if !currentWords.isEmpty, let start = currentStart, let speaker = currentSpeaker {
                    let text = currentWords.map { $0.word }.joined()
                    let end = currentWords.last?.end ?? start
                    result.append(SkribeSegment(
                        start: start,
                        end: end,
                        speaker: "SPEAKER_\(speaker)",
                        text: text.trimmingCharacters(in: .whitespaces)
                    ))
                }

            } else {
                // Fallback: segment-level alignment (old behavior)
                let text = segment.text.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !text.isEmpty else { continue }

                var bestSpeakerId: Int?
                var bestOverlap: Double = 0

                for ds in diarSegments {
                    let overlap = max(0, min(segment.end, ds.end) - max(segment.start, ds.start))
                    if overlap > bestOverlap {
                        bestOverlap = overlap
                        bestSpeakerId = ds.speakerId
                    }
                }

                if bestSpeakerId == nil, !diarSegments.isEmpty {
                    let mid = (segment.start + segment.end) / 2
                    bestSpeakerId = findSpeakerAtTime(mid, in: diarSegments)
                }

                result.append(SkribeSegment(
                    start: segment.start,
                    end: segment.end,
                    speaker: "SPEAKER_\(bestSpeakerId ?? 0)",
                    text: text
                ))
            }
        }

        return result
    }

    /// Find which speaker is active at a specific timestamp
    private static func findSpeakerAtTime(_ timestamp: Double, in segments: [DiarizationService.SpeakerSegmentResult]) -> Int {
        // Find exact match (timestamp falls within segment)
        for segment in segments {
            if timestamp >= segment.start && timestamp <= segment.end {
                return segment.speakerId
            }
        }

        // Fallback: find closest segment by midpoint
        guard !segments.isEmpty else { return 0 }
        let closest = segments.min(by: { seg1, seg2 in
            let mid1 = (seg1.start + seg1.end) / 2
            let mid2 = (seg2.start + seg2.end) / 2
            return abs(timestamp - mid1) < abs(timestamp - mid2)
        })
        return closest?.speakerId ?? 0
    }

    private static func renameSpeakers(segments: [SkribeSegment]) -> [SkribeSegment] {
        var speakerMap: [String: String] = [:]
        var counter = 1

        return segments.map { seg in
            if speakerMap[seg.speaker] == nil {
                speakerMap[seg.speaker] = "Pessoa \(counter)"
                counter += 1
            }
            return SkribeSegment(
                start: seg.start,
                end: seg.end,
                speaker: speakerMap[seg.speaker]!,
                text: seg.text
            )
        }
    }
}

File: linux/Skribe/Services/TranscriptionPipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Avalonia.Threading;
using Skribe.Models;

namespace Skribe.Services;

public sealed class TranscriptionPipeline
{
    public async Task ProcessAsync(string filePath, AppState state, ModelManager models)
    {
        Run(() =>
        {
            state.StartProcessing();
            state.FileName = System.IO.Path.GetFileName(filePath);
        });

        string? wavPath = null;
        try
        {
            Run(() => { state.CurrentPhase = ProcessingPhase.Preparing; state.ProgressMessage = "Convertendo áudio..."; });
            wavPath = await AudioProcessor.ConvertToMono16kHzAsync(filePath);

            var duration = await AudioProcessor.GetDurationAsync(wavPath);
            Run(() =>
            {
                state.AudioDuration = duration;
                state.ProgressMessage = $"Áudio preparado ({(int)duration / 60}min {(int)(duration % 60)}s)";
            });

            Run(() => { state.CurrentPhase = ProcessingPhase.Transcribing; state.ProgressMessage = "Analisando áudio..."; });
            var samples = await AudioProcessor.LoadAudioAsFloatArrayAsync(wavPath);

            var voiceSegs = VADProcessor.DetectVoiceActivity(samples);
            var voiceDur = voiceSegs.Sum(v => v.EndTime - v.StartTime);
            if (voiceDur < 1.0)
                throw new SkribeException("Nenhuma fala detectada no áudio");

            Run(() => state.ProgressMessage = $"Voz detectada: {(int)voiceDur}s de fala em {(int)duration}s totais");

            var transcribeTask = models.TranscriptionService.TranscribeAsync(wavPath);
            var diarizeTask = models.DiarizationService.DiarizeAsync(samples);

            Run(() => { state.CurrentPhase = ProcessingPhase.Diarizing; state.ProgressMessage = "Transcrevendo e identificando vozes em paralelo..."; });

            await Task.WhenAll(transcribeTask, diarizeTask);
            var whisperSegments = await transcribeTask;
            var diarSegments = await diarizeTask;

            Run(() => { state.CurrentPhase = ProcessingPhase.Merging; state.ProgressMessage = "Combinando transcrição com vozes..."; });
            var merged = AssignSpeakersWordLevel(whisperSegments, diarSegments);
            var renamed = RenameSpeakers(merged);

            Run(() => state.FinishProcessing(renamed, System.IO.Path.GetFileName(filePath)));
        }
        catch (Exception ex)
        {
            Run(() => state.FailProcessing(ex.Message));
        }
        finally
        {
            if (wavPath != null) AudioProcessor.Cleanup(wavPath);
        }
    }

    private static void Run(Action action)
    {
        if (Dispatcher.UIThread.CheckAccess()) action();
        else Dispatcher.UIThread.Post(action);
    }

    private static List<SkribeSegment> AssignSpeakersWordLevel(
        IReadOnlyList<TranscriptionService.SegmentWithWords> whisperSegs,
        IReadOnlyList<DiarizationService.SpeakerSegmentResult> diarSegs)
    {
        var result = new List<SkribeSegment>();

        foreach (var seg in whisperSegs)
        {
            if (seg.Words.Count > 0)
            {
                int? curSpeaker = null;
                var curWords = new List<TranscriptionService.WordTiming>();
                double? curStart = null;

                foreach (var w in seg.Words)
                {
                    var mid = (w.Start + w.End) / 2.0;
                    var spk = FindSpeakerAt(mid, diarSegs);
                    if (spk != curSpeaker)
                    {
                        if (curWords.Count > 0 && curStart is double s && curSpeaker is int sp)
                        {
                            result.Add(new SkribeSegment(s, curWords[^1].End, $"SPEAKER_{sp}",
                                string.Concat(curWords.Select(x => x.Word)).Trim()));
                        }
                        curSpeaker = spk;
                        curWords = new List<TranscriptionService.WordTiming> { w };
                        curStart = w.Start;
                    }
                    else curWords.Add(w);
                }
                if (curWords.Count > 0 && curStart is double s2 && curSpeaker is int sp2)
                {
                    result.Add(new SkribeSegment(s2, curWords[^1].End, $"SPEAKER_{sp2}",
                        string.Concat(curWords.Select(x => x.Word)).Trim()));
                }
            }
            else
            {
                var text = seg.Text.Trim();
                if (string.IsNullOrEmpty(text)) continue;
                int? bestSpeaker = null;
                double bestOverlap = 0;
                foreach (var d in diarSegs)
                {
                    var overlap = Math.Max(0, Math.Min(seg.End, d.End) - Math.Max(seg.Start, d.Start));
                    if (overlap > bestOverlap) { bestOverlap = overlap; bestSpeaker = d.SpeakerId; }
                }
                if (bestSpeaker is null && diarSegs.Count > 0)
                    bestSpeaker = FindSpeakerAt((seg.Start + seg.End) / 2, diarSegs);
                result.Add(new SkribeSegment(seg.Start, seg.End, $"SPEAKER_{bestSpeaker ?? 0}", text));
            }
        }

        return result;
    }

    private static int FindSpeakerAt(double t, IReadOnlyList<DiarizationService.SpeakerSegmentResult> segs)
    {
        foreach (var s in segs)
            if (t >= s.Start && t <= s.End) return s.SpeakerId;
        if (segs.Count == 0) return 0;
        var closest = segs.OrderBy(s => Math.Abs(t - (s.Start + s.End) / 2)).First();
        return closest.SpeakerId;
    }

    private static List<SkribeSegment> RenameSpeakers(IReadOnlyList<SkribeSegment> segs)
    {
        var map = new Dictionary<string, string>();
        int counter = 1;
        var output = new List<SkribeSegment>(segs.Count);
        foreach (var s in segs)
        {
            if (!map.ContainsKey(s.Speaker)) map[s.Speaker] = $"Pessoa {counter++}";
            output.Add(new SkribeSegment(s.Start, s.End, map[s.Speaker], s.Text));
        }
        return output;
    }
}
